examples/custom/01-parameter-passing/cf_demo1.c

```diff
@@ -0,0 +1,110 @@
+#include "pf_all.h"      /* lots of stuff */
+#include "cf_helpers.h"  /* to_C_string */
+#include<errno.h>        /* errno         */
+#include<limits.h>       /* PATH_MAX */
+#include<stdlib.h>       /* malloc */
+#include<stdio.h>        /* asprintf, sprintf */
+#include<string.h>       /* strerror */
+#include<sys/stat.h>     /* struct stat, stat */
+
+/* 
+ * put forward declarations here if necessary
+*/
+
+
+/****************************************************************
+** Step 1: Put your own special glue routines here
+**     or link them in from another file or library.
+****************************************************************/
+
+/* exported functions */
+
+static cell_t f4711( cell_t Val )
+{/* a quick way to check that custom words are available
+ */
+    return 11 + 47*Val;
+}
+
+static cell_t FileInfo( cell_t path_caddr, cell_t path_len ) {
+	char* path = to_C_string( path_caddr, path_len );
+	struct stat info;
+	const char* fmtErr  = "error{ id=%i, desc='%s', path='%s' }";
+	const char* fmtDir  = "directory{ path='%s' }";
+	const char* fmtFile = "file{ size=%i, path='%s' }";
+	char* result;
+	/* MSYS/Cygwin may warn than asprintf() is not defined but compile and run just fine :-/ */
+	if( stat(path, &info) == -1 )
+		asprintf( &result, fmtErr, errno, strerror(errno), path );
```

Comment:
This does not go through the Forth output path so it will not work over a serial port.

Demonstrating custom C calling does not require calling complex host functions.
It could just be a function that modifies a string.
The user will just replace these custom functions so the simpler the better.

Reply:
> This does not go through the Forth output path so it will not work over a serial port.

Just to clarify: asprintf() allocates a buffer of correct size into which it sprintf's the formatted output.
Which then gets returned. 
I'm not aware of bypassing any PForth functionality here.
 
> Demonstrating custom C calling does not require calling complex host functions. It could just be a function that modifies a string. The user will just replace these custom functions so the simpler the better.

That's what is done here.
Creating a new string (of previously unknown length) requires either hairy assumptions about maximal length or allocating a new buffer. All the fun of programming C.